A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js

```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+class A32NX_FlightPhase_PreFlight {
+    constructor() {
+        this.nextFmgcFlightPhase = FmgcFlightPhases.TAKEOFF;
+    }
+
+    init(_fmc) {
+        _fmc.climbTransitionGroundAltitude = null;
+    }
+
+    check(_fmc) {
+        // Simplane.getAltitudeAboveGround() > 1500; is used to skip flight phase in case ac reloaded midair
+        return Simplane.getEngineThrottleMode(0) >= ThrottleMode.FLEX_MCT || Simplane.getEngineThrottleMode(1) >= ThrottleMode.FLEX_MCT || Simplane.getAltitudeAboveGround() > 1500;
+    }
+}
+
+class A32NX_FlightPhase_TakeOff {
+    constructor() {
+        this.nextFmgcFlightPhase = FmgcFlightPhases.CLIMB;
+    }
+
+    init(_fmc) {
+        this.accelerationAltitudeMsl = (_fmc.accelerationAltitude || _fmc.thrustReductionAltitude);
+
+        if (!this.accelerationAltitudeMsl) {
+            if (!_fmc.climbTransitionGroundAltitude) {
+                const origin = _fmc.flightPlanManager.getOrigin();
+                if (origin) {
+                    _fmc.climbTransitionGroundAltitude = origin.altitudeinFP;
+                }
+
+                if (!_fmc.climbTransitionGroundAltitude) {
+                    _fmc.climbTransitionGroundAltitude = (parseInt(SimVar.GetSimVarValue("GROUND ALTITUDE", "feet")) || 0);
+                }
+            }
+
+            this.accelerationAltitudeMsl = _fmc.climbTransitionGroundAltitude + parseInt(NXDataStore.get("CONFIG_ACCEL_ALT", "1500"));
+        }
+    }
+
+    check(_fmc) {
+        return Simplane.getAltitude() > this.accelerationAltitudeMsl;
+    }
+}
+
+class A32NX_FlightPhase_Climb {
+    constructor() {
+        this.nextFmgcFlightPhase = FmgcFlightPhases.CRUISE;
+    }
+
+    init(_fmc) {
+    }
+
+    check(_fmc) {
+        return Math.round(Simplane.getAltitude() / 100) >= _fmc.cruiseFlightLevel;
+    }
+}
+
+class A32NX_FlightPhase_Cruise {
+    constructor() {
+    }
+
+    init(_fmc) {
+    }
+
+    check(_fmc) {
+        return false;
+    }
+}
+
+class A32NX_FlightPhase_Descent {
+    constructor() {
+    }
+
+    init(_fmc) {
+        this.nextFmgcFlightPhase = FmgcFlightPhases.APPROACH;
+    }
+
+    check(_fmc) {
+        const fl = Math.round(Simplane.getAltitude() / 100);
+        const fcuSelFl = Simplane.getAutoPilotDisplayedAltitudeLockValue("feet") / 100;
+        if (fl === _fmc.cruiseFlightLevel && fcuSelFl === fl) {
+            this.nextFmgcFlightPhase = FmgcFlightPhases.CRUISE;
+            return true;
+        }
+
+        if (_fmc.flightPlanManager.decelWaypoint) {
+            const lat = SimVar.GetSimVarValue("PLANE LATITUDE", "degree latitude");
+            const long = SimVar.GetSimVarValue("PLANE LONGITUDE", "degree longitude");
+            const planeLla = new LatLongAlt(lat, long);
+            const dist = Avionics.Utils.computeGreatCircleDistance(_fmc.flightPlanManager.decelWaypoint.infos.coordinates, planeLla);
+            if (dist < 3) {
+                _fmc.flightPlanManager._decelReached = true;
```

Comment:
It's a bit naughty to write another classes members like this.

Reply:
yes I don't like it either, eventually this won't be in the flight phase manager, but in the fmgc somewhere itself. possibly getting some inputs from "flight plan manager" or so to call our flight phase change method (which exists for exactly these type of actions)

Reply:
That's cool. Maybe a TODO comment that will highlight in most peoples editors.

Reply:
Added todo